### PR TITLE
ERC20 rate limiter

### PR DESCRIPTION
**Pull request by viraj124**
- closes https://github.com/FuelLabs/fuel-bridge/issues/210
- adds rate limit feature that can be set individually for each erc20 token that the bridge supports

### CODE REVIEW

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol [hunk 1]
```diff
@@ -66,10 +71,22 @@ contract FuelERC20GatewayV4 is
...
-    mapping(address => uint256) internal _depositLimits;
-    mapping(address => uint256) internal _decimalsCache;
-    mapping(bytes32 => bool) internal _isBridge;
+    mapping(address => uint256) _deposits;
+    mapping(address => uint256) _depositLimits;
+    mapping(address => uint256) _decimalsCache;
+    mapping(bytes32 => bool) _isBridge;
+
+    /// @notice Amounts already withdrawn this period for each token.
+    mapping(address => uint256) public rateLimitDuration;
```
Comment (DefiCake) [inline packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:80]: let's the take the chance to implement https://github.com/FuelLabs/fuel-bridge/issues/173

until mainnet launch we won't care if the erc20 gateway breaks backwards compatibility in storage

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol [hunk 2]
```diff
@@ -121,6 +140,57 @@ contract FuelERC20GatewayV4 is
...
     }
 
+    /**
+     * @notice Intitializing rate limit params for each token. 
+     * @param _token The token address to set rate limit params for.
+     * @param _limitAmount The amount to reset the limit to.
+     * @param _rateLimitDuration rate limit duration.
+     */
+    function initializeRateLimit(address _token, uint256 _limitAmount, uint256 _rateLimitDuration) external onlyRole(SET_RATE_LIMITER_ROLE) {
+        if (currentPeriodEnd[_token] != 0) revert RateLimitAlreadySet();
```
Comment (DefiCake) [inline packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:150]: Why isn't it possible to change the duration after it has been set?

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol [hunk 3]
```diff
@@ -226,6 +296,9 @@ contract FuelERC20GatewayV4 is
         uint8 decimals = _getTokenDecimals(tokenAddress);
         uint256 amount = _adjustWithdrawalDecimals(decimals, l2BurntAmount);
 
+        // rate limit check
+        _addWithdrawnAmount(tokenAddress, amount);
```
Comment (DefiCake) [inline packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:300]: Wouldn't we achieve significant gas savings if we added a check for the existence of a rate limiter?

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol [hunk 4]
```diff
@@ -121,6 +140,57 @@ contract FuelERC20GatewayV4 is
...
+        currentPeriodEnd[_token] = block.timestamp + _rateLimitDuration;
+        limitAmount[_token] = _limitAmount;
+    }
+
+    /**
+     * @notice Resets the rate limit amount.
+     * @param _token The token address to set rate limit for.
+     * @param _amount The amount to reset the limit to.
+     */
+    function resetRateLimitAmount(address _token, uint256 _amount) external onlyRole(SET_RATE_LIMITER_ROLE) {
```
Comment (DefiCake) [inline packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:162]: Shouldn't we have events?

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol (hunk 1 above)
Comment (DefiCake) [reply packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:80]: Ignore the above comment, we are finally going to go with the old storage

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol [hunk 5]
```diff
@@ -66,10 +74,22 @@ contract FuelERC20GatewayV4 is
...
     bytes32 public assetIssuerId;
 
-    mapping(address => uint256) internal _deposits;
-    mapping(address => uint256) internal _depositLimits;
-    mapping(address => uint256) internal _decimalsCache;
-    mapping(bytes32 => bool) internal _isBridge;
+    mapping(address => uint256) _deposits;
+    mapping(address => uint256) _depositLimits;
+    mapping(address => uint256) _decimalsCache;
+    mapping(bytes32 => bool) _isBridge;
```
Comment (PraneshASP) [inline packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:80]: nit: good to add visibility as `private`

Comment (PraneshASP) [reply packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:80]: also we need to decide the naming convention. should we use `_` prefix for private vars or only for params and local vars?

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol (hunk 2 above)
Comment (PraneshASP) [reply packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:153]: so the workflow would be to call initialzeRateLimit() first then to change the config `resetRateLimit()` must be used?

also good to emit an event here. can be generic as `RateLimitUpdated()` which can be used in both methods.

**Review by PraneshASP — COMMENTED** [review]
few nits

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol (hunk 3 above)
Comment (DefiCake) [reply packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:311]: We cannot do that, as we cannot initialize a rate limiter for every token that will ever exist. By default we must use no rate limiter

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol (hunk 2 above)
Comment (DefiCake) [reply packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:153]: Please make it dynamic. The reasons why it is static in FuelMessagePortal is that we can change it via new deployment (it is a constant / immutable), but here we cannot leverage that advantage, as we need an storage variable for each token. Thus it makes no sense to have that limitation.

File: packages/solidity-contracts/test/erc721Gateway.test.ts [hunk 6]
```diff
@@ -32,7 +32,6 @@ import {
   TIME_TO_FINALIZE,
   COMMIT_COOLDOWN,
 } from './utils';
-
```
Comment (DefiCake) [inline packages/solidity-contracts/test/erc721Gateway.test.ts:35]: ?

File: packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol [hunk 7]
```diff
@@ -66,10 +74,22 @@ contract FuelERC20GatewayV4 is
     bool public whitelistRequired;
     bytes32 public assetIssuerId;
 
-    mapping(address => uint256) internal _deposits;
-    mapping(address => uint256) internal _depositLimits;
-    mapping(address => uint256) internal _decimalsCache;
-    mapping(bytes32 => bool) internal _isBridge;
+    mapping(address => uint256) private _deposits;
```
Comment (DefiCake) [inline packages/solidity-contracts/contracts/messaging/gateway/FuelERC20Gateway/FuelERC20GatewayV4.sol:77]: Why make them private?

**Review by DefiCake — APPROVED** [review]
lgtm

**Pull request merged**